### PR TITLE
[Calyx] Add group-invariant code motion pass

### PR DESCRIPTION
**Pull request by mortbopet**
The assumption here is that any operation which is not a special group operation (`calyx.group_done,calyx.group_go,calyx.assign`) is some form of combinational computation that can be lifted to the wire scope.

The pattern is added as a separate pass, but could be added as a canonicalization of `GroupOp`s.
I'd be in favor of adding this as a canonicalization of `GroupOp`s. GICM does not mean that we suddenly get unwanted sharing of `comb` operation results (e.g. as seen in the test, an identical `comb` operation is emitted twice in the wire scope - CSE is not part of `comb` canonicalizers, which i presume is due to fan-out reasons).
The downside to doing GICM during canonicalization is that it might obscure the IR a slight bit, since `comb` operations are no longer inside the groups where they were initially placed during either human IR writing, or when some other pass generated the IR... (but i guess that could also be said for other canonicalization patterns!).

Let me know what you think.

### CODE REVIEW

File: lib/Dialect/Calyx/Transforms/GICM.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+//===- GICM.cpp - Group-invariant code motion pass --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass performs GICM (group-invariant code motion) of operations which are
+// deemed to be invariant of the group which they are placed in. In practice,
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Transforms/GICM.cpp:10]:
```suggestion
// deemed to be invariant of the group in which they are placed. In practice,
```

**Review by cgyurgyik — APPROVED** [review]
This seems reasonable, as it is an artifact of working in MLIR.

**Comment by cgyurgyik** [comment]
> IMO human IR writing is at best a secondary concern.

+1, Calyx IR was meant to be generated by a front-end.

**Comment by mortbopet** [comment]
The specific issue that I'm trying to solve with this pass is when doing e.g. control compilation, we're inspecting the `group_done` source and guard values of various groups. If these are defined by `comb` expressions within a group, then these values cannot be referenced from other groups due to lack of dominance.  

That could also be an argument for canonicalization - making it so that no operation within a group ever defines a value (this could even extend to the `calyx.group_go` operation if it is made to reference a group instead of next it inside a group).

This restriction (and GICM) fits with my mental model of groups - they never define anything structural, they strictly describe wiring between existing structures.

**Comment by mikeurbach** [comment]
That all sounds good, but please put it in the documentation of the pass :) It is starting to sound like this is a good canonicalization, but again I think this should land first and we should then define a cost model before making such a move.

**Pull request merged**